### ISSUE TITLE
Throw uncaught  Exception

### ISSUE THREAD

**Issue by socketman2016** [jakubkulhan/bunny#36]
I ask this question https://github.com/reactphp/promise/issues/63
In last reply

> The code is probably wrapped into another promise which catches the exception. You must ensure that you call done() on any promise which you don't return for comsumption, eg. from a function. See done() vs. then().

I what to know where is the wrapper promise? 
```php
$request = my request object
$this->_channel->queueDeclare($request->getQueueName(), false, $request::PERSISTENT)->then(function () use ($request) {
                $this->publish($request);
            }, function (ClientException $e) {
                throw new \Exception();

            })->done();
```

Why react/promise/src/FulfilledPromise.php:26 catch my exception?

**Comment by jsor** [comment]
I've opened #37 which should fix this.

**Comment by jakubkulhan** [comment]
@socketman2016 I've merged @jsor's PR, can you try again with version from `master`?

**Comment by socketman2016** [comment]
I get with version from ```master``` , but not works?!! :frowning: 

```php
$request = my request object
$this->_channel->queueDeclare($request->getQueueName(), false, $request::PERSISTENT)->then(function () use ($request) {
                $this->publish($request);
            }, function (ClientException $e) {
               var_dump($e->getMessage());
                throw $e;

            })->done();
```

```var_dump``` show me
> string(118) "PRECONDITION_FAILED - inequivalent arg 'durable' for queue 'test' in vhost '/': received 'true' but current is 'false'"

but ```throw $e;``` catched!!

**Comment by jakubkulhan** [comment]
You must have already declared queue with the same name (`test`), however, with different arguments (as message says, the problem is different argument `durable`). RabbitMQ doesn't allow to redeclare queue with different arguments. You first have to delete given queue (either manually through RabbitMQ management GUI, or by calling `queueDelete`) and then declare it again.

**Comment by socketman2016** [comment]
I know , it is only a test , I just want to create an error case

I want to know , why is this error case I cannot throw an uncaught exception?
I have an uncaught exception handler that log any error in my system

**Comment by jsor** [comment]
Maybe you've wrapped the above code in another Promise yourself. I've setup a minimal reproducible test case and it throws the exception.

```php
<?php

require __DIR__ . '/vendor/autoload.php';

$bunny = new Bunny\Async\Client();

$bunny
    ->connect()
    ->then(function (Bunny\Async\Client $bunny) {
        return $bunny->channel();
    })
    ->then(function (Bunny\Channel $channel) {
        return $channel->queueDeclare('issue36', false, true);
    })
    ->done()
;

$bunny->run();
```

When i change the third argument of `queueDeclare()`to `false` and run it again, it throws the exception.
<details>
 <summary>Exception</summary>
```
PHP Fatal error:  Uncaught Bunny\Exception\ClientException: PRECONDITION_FAILED - inequivalent arg 'durable' for queue 'issue36' in vhost '/': received 'false' but current is 'true' in /Users/jsor/Code/bunny/src/Bunny/ClientMethods.php:1259
Stack trace:
#0 /Users/jsor/Code/bunny/vendor/react/promise/src/FulfilledPromise.php(39): Bunny\AbstractClient->Bunny\{closure}(true)
#1 /Users/jsor/Code/bunny/vendor/react/promise/src/Promise.php(51): React\Promise\FulfilledPromise->done(Object(Closure), NULL)
#2 /Users/jsor/Code/bunny/vendor/react/promise/src/Promise.php(156): React\Promise\Promise->React\Promise\{closure}(Object(React\Promise\FulfilledPromise))
#3 /Users/jsor/Code/bunny/vendor/react/promise/src/Promise.php(128): React\Promise\Promise->settle(Object(React\Promise\FulfilledPromise))
#4 /Users/jsor/Code/bunny/vendor/react/promise/src/Promise.php(169): React\Promise\Promise->resolve(true)
#5 [internal function]: React\Promise\Promise->React\Promise\{closure}(true)
#6 /Users/jsor/Code/bunny/vendor/react/promise/src/Deferred.p in /Users/jsor/Code/bunny/src/Bunny/ClientMethods.php on line 1259

Fatal error: Uncaught Bunny\Exception\ClientException: PRECONDITION_FAILED - inequivalent arg 'durable' for queue 'issue36' in vhost '/': received 'false' but current is 'true' in /Users/jsor/Code/bunny/src/Bunny/ClientMethods.php on line 1259

Bunny\Exception\ClientException: PRECONDITION_FAILED - inequivalent arg 'durable' for queue 'issue36' in vhost '/': received 'false' but current is 'true' in /Users/jsor/Code/bunny/src/Bunny/ClientMethods.php on line 1259

Call Stack:
    0.0003     356808   1. {main}() /Users/jsor/Code/bunny/test.php:0
    0.0107    1953352   2. Bunny\Async\Client->run() /Users/jsor/Code/bunny/test.php:18
    0.0107    1953352   3. React\EventLoop\StreamSelectLoop->run() /Users/jsor/Code/bunny/src/Bunny/Async/Client.php:117
    0.0170    2193880   4. React\EventLoop\StreamSelectLoop->waitForStreamActivity() /Users/jsor/Code/bunny/vendor/react/event-loop/StreamSelectLoop.php:198
    0.0170    2194360   5. call_user_func:{/Users/jsor/Code/bunny/vendor/react/event-loop/StreamSelectLoop.php:232}() /Users/jsor/Code/bunny/vendor/react/event-loop/StreamSelectLoop.php:232
    0.0170    2194360   6. Bunny\Async\Client->Bunny\Async\{closure}() /Users/jsor/Code/bunny/vendor/react/event-loop/StreamSelectLoop.php:232
    0.0178    2230136   7. React\Promise\Deferred->reject() /Users/jsor/Code/bunny/src/Bunny/Async/Client.php:170
    0.0178    2230136   8. call_user_func:{/Users/jsor/Code/bunny/vendor/react/promise/src/Deferred.php:42}() /Users/jsor/Code/bunny/vendor/react/promise/src/Deferred.php:42
    0.0178    2230136   9. React\Promise\Promise->React\Promise\{closure}() /Users/jsor/Code/bunny/vendor/react/promise/src/Deferred.php:42
    0.0178    2230136  10. React\Promise\Promise->reject() /Users/jsor/Code/bunny/vendor/react/promise/src/Promise.php:172
    0.0178    2230192  11. React\Promise\Promise->settle() /Users/jsor/Code/bunny/vendor/react/promise/src/Promise.php:137
    0.0178    2230192  12. React\Promise\Promise->React\Promise\{closure}() /Users/jsor/Code/bunny/vendor/react/promise/src/Promise.php:156
    0.0178    2230192  13. React\Promise\RejectedPromise->done() /Users/jsor/Code/bunny/vendor/react/promise/src/Promise.php:51
```
</details>

**Comment by socketman2016** [comment]
Oh my God!!! What happens?!!
I create a new test.php file and copy your code
first time I run it ```issue36``` created durable , second time I change the third argument of ```queueDeclare()``` to ```false``` and  it throws the uncaught  exception and every thing is ok

but third time when third argument of ```queueDeclare()``` is ```false``` nothing happen! no exception thrown?! please try your self 

I run test.php again and again but no exception thrown

**Comment by jsor** [comment]
I cannot reproduct your behaviour. I throws everytime i run the script with different arguments from the first run.

**Comment by socketman2016** [comment]
Where is my problem?
I use ```composer update bunny/bunny``` to get last version , compose told me updated to commit ```f81da29e8a279af2b9af80fa041b84497a26ddf8``` Also I check vendor folder and I see changed then to done

Now I change your code to this , (I have a durable issue36 queue so this is error case)
```php
<?php

require __DIR__ . '/vendor/autoload.php';

$bunny = new Bunny\Async\Client();

$bunny
    ->connect()
    ->then(function (Bunny\Async\Client $bunny) {
        var_dump('Connected');
        return $bunny->channel();
    })
    ->then(function (Bunny\Channel $channel) {
        var_dump('Channel is gotten');
        return $channel->queueDeclare('issue36', false, false);
    })
    ->done(function (){
        var_dump('Done');
    },function (\Bunny\Exception\ClientException $e){
        var_dump($e->getMessage());
        throw $e;
    })
;

$bunny->run();
```

every this result is
>string(9) "Connected"
string(17) "Channel is gotten"
string(121) "PRECONDITION_FAILED - inequivalent arg 'durable' for queue 'issue36' in vhost '/': received 'false' but current is 'true


And no uncaught Exception

Why? what you think?!

**Comment by socketman2016** [comment]
@jsor , Can you send me zip of your working code
I am confusing and cannot understand where is my problem

**Comment by jsor** [comment]
I found the reason why it was working for me: I had react/promise v2.2.0 installed because i tested my PR with `composer update --prefer-lowest` and in this version the exception is thrown due to a bug we fixed in react/promise v2.2.1.

I'm still investigating this behaviour and i'm still not sure where the problem is. I first suspected it to be a bug in react/promise but i couldn't reproduce it as a test case.

I think now the exception must be trapped somwhere in the bunny code. There are still a lot of place where promise creating methods are called without calling `done()`on the promise.

One example is [`flushWriteBuffer()`](https://github.com/jakubkulhan/bunny/blob/master/src/Bunny/Async/Client.php#L149) which is called often without returning and without calling `done()` on the promise ([example](https://github.com/jakubkulhan/bunny/blob/master/src/Bunny/ClientMethods.php#L1239)). The problem is, that it only returns a promise for the async client.

To be honest, i'm not sure how to fix this atm :worried:

**Comment by socketman2016** [comment]
@jakubkulhan Please reopen issue
@jsor before ```var_dump($e->getMessage());``` in my last code , I put ```echo debug_print_backtrace();```

the result maybe help us to find problem
>string(9) "Connected"
string(17) "Channel is gotten"
* {closure}() called at [vendor\react\promise\src\RejectedPromise.php:39]
* React\Promise\RejectedPromise->done() called at [vendor\react\promise\src\Promise.php:51]
* React\Promise\Promise->React\Promise\{closure}() called at [vendor\react\promise\src\Promise.php:164]
* React\Promise\Promise->settle() called at [vendor\react\promise\src\Promise.php:142]
* React\Promise\Promise->reject() called at [vendor\react\promise\src\Promise.php:185]
* React\Promise\Promise->React\Promise\{closure}() called at [vendor\react\promise\src\RejectedPromise.php:39]
* React\Promise\RejectedPromise->done() called at [vendor\react\promise\src\Promise.php:51]
* React\Promise\Promise->React\Promise\{closure}() called at [vendor\react\promise\src\Promise.php:164]
* React\Promise\Promise->settle() called at [vendor\react\promise\src\Promise.php:142]
* React\Promise\Promise->reject() called at [vendor\react\promise\src\Promise.php:185]
* React\Promise\Promise->React\Promise\{closure}()
* call_user_func() called at [vendor\react\promise\src\Deferred.php:42]
* React\Promise\Deferred->reject() called at [vendor\bunny\bunny\src\Bunny\ClientMethods.php:1259]
* Bunny\AbstractClient->Bunny\{closure}() called at [vendor\react\promise\src\FulfilledPromise.php:39]
* React\Promise\FulfilledPromise->done() called at [vendor\react\promise\src\Promise.php:51]
* React\Promise\Promise->React\Promise\{closure}() called at [vendor\react\promise\src\Promise.php:164]
* React\Promise\Promise->settle() called at [vendor\react\promise\src\Promise.php:133]
* React\Promise\Promise->resolve() called at [vendor\react\promise\src\Promise.php:182]
* React\Promise\Promise->React\Promise\{closure}()
* call_user_func() called at [vendor\react\promise\src\Deferred.php:35]
* React\Promise\Deferred->resolve() called at [vendor\bunny\bunny\src\Bunny\Async\Client.php:164]
* Bunny\Async\Client->Bunny\Async\{closure}()
* call_user_func() called at [vendor\react\event-loop\src\StreamSelectLoop.php:240]
* React\EventLoop\StreamSelectLoop->waitForStreamActivity() called at [vendor\react\event-loop\src\StreamSelectLoop.php:201]
* React\EventLoop\StreamSelectLoop->run() called at [vendor\bunny\bunny\src\Bunny\Async\Client.php:117]
* Bunny\Async\Client->run() called at [a.php:31]

>string(121) "PRECONDITION_FAILED - inequivalent arg 'durable' for queue 'issue36' in vhost '/': received 'false' but current is 'true'"
